test/common/network/dns_impl_test.cc

```diff
@@ -107,31 +125,81 @@ class TestDnsServerQuery {
         const std::list<std::string>* ips = nullptr;
         // We only expect resources of type A or AAAA.
         const int q_type = DNS_QUESTION_TYPE(question + name_len);
+        std::string cname;
+        // check if we have a cname. If so, we will need to send a response element with the cname
+        // and lookup the ips of the cname and send back those ips (if any) too
+        auto cit = parent_.cnames_.find(name);
+        if (cit != parent_.cnames_.end()) {
+          cname = cit->second;
+        }
+        const char* hostLookup = name;
+        const unsigned char* ip_question = question;
+        long ip_name_len = name_len;
+        std::string encodedCname;
+        if (cname.size() > 0) {
+          hostLookup = cname.c_str();
+          encodedCname = TestDnsServerQuery::encodeDnsName(cname);
+          ip_question = reinterpret_cast<const unsigned char*>(encodedCname.c_str());
+          ip_name_len =
+              encodedCname.size() + 1; //+1 as we need to include the final null terminator
+        }
         ASSERT_TRUE(q_type == T_A || q_type == T_AAAA);
         if (q_type == T_A) {
-          auto it = parent_.hosts_A_.find(name);
+          auto it = parent_.hosts_A_.find(hostLookup);
           if (it != parent_.hosts_A_.end()) {
             ips = &it->second;
           }
         } else {
-          auto it = parent_.hosts_AAAA_.find(name);
+          auto it = parent_.hosts_AAAA_.find(hostLookup);
           if (it != parent_.hosts_AAAA_.end()) {
             ips = &it->second;
           }
         }
         ares_free_string(name);
 
+        int answer_size = ips != nullptr ? ips->size() : 0;
+        answer_size += encodedCname.size() > 0 ? 1 : 0;
+
         // The response begins with the intial part of the request
         // (including the question section).
         const size_t response_base_len = HFIXEDSZ + name_len + QFIXEDSZ;
         unsigned char response_base[response_base_len];
         memcpy(response_base, request, response_base_len);
         DNS_HEADER_SET_QR(response_base, 1);
         DNS_HEADER_SET_AA(response_base, 0);
-        DNS_HEADER_SET_RCODE(response_base, ips != nullptr ? NOERROR : NXDOMAIN);
-        DNS_HEADER_SET_ANCOUNT(response_base, ips != nullptr ? ips->size() : 0);
+        DNS_HEADER_SET_RCODE(response_base, answer_size > 0 ? NOERROR : NXDOMAIN);
+        DNS_HEADER_SET_ANCOUNT(response_base, answer_size);
         DNS_HEADER_SET_NSCOUNT(response_base, 0);
         DNS_HEADER_SET_ARCOUNT(response_base, 0);
+        // Total response size will be computed according to cname response size + ip response sizes
```

Comment:
Optional - this function is getting reaaally long.  If there's anything else you can easily factor into a utility helper that'd be nice. Optional because it's test code and you've done a great job with comments and variable names.

Reply:
I'll pass on this. I agree the function is getting long. But, I am not too sure that refactoring will make this easier to understand.